### DISCUSSION TITLE
#3610后续

### DISCUSSION THREAD

**Discussion by MoYuan-CN** [Meloong-Git/PCL#3618] [讨论]
*No Response*

**Comment by MoYuan-CN** [comment]
@wuliaodexiaoluo

你拿错方法了吧，405

**Reply by wuliaodexiaoluo** [reply]
不是 put 吗？

**Reply by MoYuan-CN** [reply]
Push试试看（我没看文档）
反正 405 代表方法不被允许

**Reply by wuliaodexiaoluo** [reply]
az
Requests 没有这个方法
Requuests 只有 GET POST HEAD DELETE PUT 
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/1dca675e-4844-4e41-9a61-c5df5df9f28a)
而且文档写了是用 PUT

**Reply by wuliaodexiaoluo** [reply]
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/4c696a9d-9745-4350-873e-17b64264f6cf)
你看哪里有这个方法

**Reply by MoYuan-CN** [reply]
啊呸，是 POST
so 为啥会405呢

**Reply by wuliaodexiaoluo** [reply]
POST 结果
<function put at 0x038FE348>
我感觉根本没传上去

**Reply by hejiehao** [reply]
![image](https://github.com/Hex-Dragon/PCL2/assets/98140773/941e11a9-c246-4b4d-a46b-c817d5353ed5)
POST https://api.minecraftservices.com/minecraft/profile/skins
上面的运行是用的 128*128 的皮肤
下面的运行是用的 64*64 的皮肤

**Reply by mc-124** [reply]
应该是mojang的api不支持，但mc的皮肤可能是根据百分比去截图的，皮肤站没做超尺寸检查，所以mc支持，皮肤站也没事，正版不行

**Reply by mc-124** [reply]
所以3610与PCL2无关

**Reply by wuliaodexiaoluo** [reply]
![image](https://github.com/Hex-Dragon/PCL2/assets/160083863/b66af6fd-2887-487a-a9f2-cebc46d5dc51)
``` Python
 headers = {
...     "Authorization":"Bearer eyJraWQiOiJhYzg0YSIsImFsZyI6IkhTMjU2In0.eyJ4dWlkIjoiMjUzNTQ2MTExNTE5MjM1MiIsImFnZyI6IkFkdWx0Iiwic3ViIjoiZWFhZWMxOWUtMGUyMy00NTAyLTk2ODYtODc5OTczYzNmZWM2IiwiYXV0aCI6IlhCT1giLCJucyI6ImRlZmF1bHQiLCJyb2xlcyI6W10sImlzcyI6ImF1dGhlbnRpY2F0aW9uIiwiZmxhZ3MiOlsidHdvZmFjdG9yYXV0aCIsIXXXXXXXtMDY3NzQ5NGE0NjQwIn0sInBsYXRmb3JtIjoiVU5LTk9XTiIsInl1aWQiOiI1NmFmZDU0N2Y3OWIyZTA1YzE0OGU4NWQ1MjNkYmE1NSIsIm5iZiI6MTcxMjMxMjMwMywiZXhwIjoxNzEyMzk4NzAzLCJpYXQiOjE3MTIzMTIzMDN9.mrx696zHmzETocIw3HkEbGq4Sf6nmCjWZzvaPEjpfCc"
... }
>>> data = {    "",    d()
...     "model":"",
...     "file":skin.read()
... }
>>> resopen = requests.put(url="https://api.mojang.com/user/profile/39a9f12c34eb44c4af090677494a4640/skin",headers=headers_2,data=data)
>>> respoen
<function put at 0x038FE348>
```
相关实现，之前也是这么搞得，Put 405 后就没成功过

**Reply by allMagicNB** [reply]
我这报错了，相关指令：
```command
curl -X POST -H "Authorization: Bearer HeyBroWhatRUThinking" -F variant=classic -F file="@皮肤.png;type=image/png" https://api.minecraftservices.com/minecraft/profile/skins
```
（命令执行的目录下有 `皮肤.png` 文件）
返回 Json：
```json
{
  "path" : "/minecraft/profile/skins",
  "errorMessage" : "Could not read image data."
}
```

**Reply by hejiehao** [reply]
```python
import requests
from requests_toolbelt import MultipartEncoder

skin = MultipartEncoder(
    fields={'variant': "slim",
            "file": ("苦力怕娘.png", open("./苦力怕娘.png", "rb"))}
)

response = requests.post("https://api.minecraftservices.com/minecraft/profile/skins", data=skin, headers={
    "Authorization": "Bearer Hey Bro,What are you thinking?Peach?",
    "Content-Type": skin.content_type
})

print(response)
print(response.text)
```

**Reply by wuliaodexiaoluo** [reply]
貌似那个 Put xxxxx 才是对的，因为这边上传成功是没有应答
POST 是更换皮肤，PUT是上传和设置
这俩接口为什么要重复？

**Reply by Silverteal** [reply]
一个从远端获取，一个从本地获取。
详见[wiki.vg](https://wiki.vg/Mojang_API#Change_Skin)

**Reply by wuliaodexiaoluo** [reply]
到底是 POST 还是 PUT
麻了 Mojang API 好麻烦

**Reply by hejiehao** [reply]
我用的 https://api.minecraftservices.com/minecraft/profile/skins

**Reply by wuliaodexiaoluo** [reply]
Mojang 到底做了多少个重复的接口
UUID 可以从 api.mojang.com 查，也可以从 api.minecraftservices.com 查
无语了真的

**Reply by Silverteal** [reply]
> resopen = requests.put(url="https://api.mojang.com/user/profile/39a9f12c34eb44c4af090677494a4640/skin",headers=headers_2,data=data)

> respoen
<function put at 0x038FE348>

难以理解的结果。尝试更新requests？

**Reply by Silverteal** [reply]
> Mojang 到底做了多少个重复的接口
UUID 可以从 api.mojang.com 查，也可以从 api.minecraftservices.com 查
无语了真的

对于新开发的应用，建议使用api.minecraftservices.com 新接口，原有的api.mojang.com 只为兼容性保留

**Reply by hejiehao** [reply]
![image](https://github.com/Hex-Dragon/PCL2/assets/98140773/d8e102cf-e3a1-43dc-ab7e-ed1cd07897e3)

**Comment by wuliaodexiaoluo** [comment]
> 难以理解的结果。尝试更新requests？

这算是 PDM 可以安装的最新版本了，再往上 3.7.4 就不支持了
